sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala

```diff
@@ -75,7 +75,7 @@ trait NamedExpression extends Expression {
    * multiple qualifiers, it is possible that there are other possible way to refer to this
    * attribute.
    */
-  def qualifiedName: String = (qualifier :+ name).mkString(".")
+  def qualifiedName: String = (qualifier :+ name).map(quoteIfNeeded).mkString(".")
```

Comment:
I was also thinking about fixing here but I thought that `toSQLId` would split the identifier anyway. Could you elaborate on why adding `quoteIfNeeded` passes the parsing in `QueryCompilationErrors.unresolvedAttributeError`?

Other than that, looks good. Thanks for adding the tests.

Reply:
Currently, `unresolvedAttributeError` calls into `toSQLId(String)`, which expects quoted parts in a single string. It parses the string while considering quotation and splits it into proper parts (`Seq[String]`). With `quoteIfNeeded` in `qualifiedName`, the parts are quoted if they contain `.`, before the parts are joined with `.`s to a single string.

I also thought about calling into `toSQLId(Seq[String])` by mapping the single candidate to `Seq(candidate)`, which fixes this as well. That looks like the least impact solution. But if `qualifiedName` needs to be fixed anyway, touching `unresolvedAttributeError` would not be needed then.